vllm/attention/backends/flashinfer.py

```diff
@@ -552,26 +537,73 @@ def _add_seq_group(
-        self.paged_kv_indices.extend(block_table[:block_table_bound])
-        self.paged_kv_indptr.append(self.paged_kv_indptr[-1] +
-                                    block_table_bound)
+            self._update_unique_kv_tensors(block_table, seq_len, common_prefix)
 
-        last_page_len = seq_len % self.block_size
+    def _update_unique_kv_tensors(self, block_table: List[int], seq_len: int,
+                                  common_prefix: List[int]) -> None:
+        """
+        Updates the unique level kv tensors
+        """
+        shared_length = len(common_prefix)
+        self.total_blocks += (len(block_table) - shared_length)
+        block_table_bound = (seq_len) // self.block_size + 1 \
+                            if seq_len % self.block_size != 0 \
+                            else (seq_len) // self.block_size
+        self.second_layer_kv_indices.extend(
+            block_table[shared_length:block_table_bound])
+        self.second_layer_kv_indptr.append(self.second_layer_kv_indptr[-1] +
+                                           (block_table_bound - shared_length))
+        # FIXME: this calculation will need to be changed
+        # once shared level is working
+        last_page_len = (seq_len) % self.block_size
         if last_page_len == 0:
             last_page_len = self.block_size
-        self.paged_kv_last_page_len.append(last_page_len)
+        self.second_layer_kv_last_page_len.append(last_page_len)
+
+    def _update_shared_kv_tensors(self, common_prefix: List[int],
+                                  batch_size: int) -> None:
+        """
+        Updates the shared level kv tensors
+        """
+        if len(common_prefix) == 0:
+            # if we don't have common prefixes, we only use the unique level
+            # so we fill the first level indices, indptr, last page len with 0s
+            # to conform with multilevel wrapper input requirements
+            self.paged_kv_indices.extend([0] * batch_size)
+            self.paged_kv_indptr.extend([0] * batch_size)
+            self.paged_kv_last_page_len.extend([0] * batch_size)
+        else:
+            # FIXME: this still doesn't work
+            self.total_blocks += len(common_prefix)
+            self.paged_kv_indices.extend(common_prefix)
+            self.paged_kv_indptr.append(self.paged_kv_indptr[-1] +
+                                        len(common_prefix))
+            self.paged_kv_last_page_len.append(self.block_size)
+
+    def get_shared_blocks_nums(
+        self,
+        inter_data_list: List["ModelInputForGPUBuilder.InterDataForSeqGroup"]
+    ) -> List[int]:
+        """
+        Returns a list of consecutive shared blocks across sequence groups
+        """
+        block_id_lists = [
+            list(data.block_tables.values()) for data in inter_data_list
+            if data.block_tables
+        ]
+
+        flattened_lists = [
+            item for sublist in block_id_lists for item in sublist
+        ]
```

Comment:
Should be more efficient:
```suggestion
        flattened_lists = []
        for data in inter_data_list:
            if data.block_tables:
                flattened_lists += data.block_tables.values()
```